### ISSUE TITLE
v1.0 rc1 design review

### ISSUE THREAD

**Issue by hartytp** [sinara-hw/Stabilizer#2]
Schematic at https://github.com/sinara-hw/Stabilizer/files/2592309/Stabilizer.PDF

Review of the main analog outputs (DAC):
- [x] Please move the 10R||10uF reference filter to between the reference and the buffer, not between the buffer and the DAC. The reason for this is that the DAC's reference input minimum resistance is only 7.5k, but very code-dependent. For Vref=4.096V, Iref=550uA, so the drop across the 10R resistor is 550uA*10R=5.5mV. 1LSB=4.096/2^16=62.5uV, so that's a very significant voltage drop, which will change a lot with the DAC code and hence kill the INL/DNL. The DAC must be driven directly from a reference buffer with very low impedance (50mOhm or so). 
- [x] the DAC data sheet recommends 10uF||100nF on the DAC Vref pin. That's going to make the reference buffer oscillate if we're not careful. What do we do about this? The LTC6655 has very good low-frequency noise, but its broadband noise is 40nV/rtHz which is too high to connect directly to the DAC. Do we choose a different reference buffer? Do we try to isolate the buffer from the capacitive load with a fb (low DC resistance)? Or do we choose a different reference and drive the DAC directly from the reference (will this require a separate reference for each DAC to keep the cross-talk acceptable?)?
- [x] Might be better to use a 10R resistor on the VDlogic pin as well as/instead of the ferret bead to isolate the digital signal supply from the analog supply. That helps to isolate them even at the 1MHz update rate (where the ferrite doesn't do much).
- [x] The DAC data sheet also recommends 10uF tantalum caps near the DAC for both power supplies. Not sure if this is really necessary.
- [x] Isn't it possible to combine analog output filtering and gain into a single stage? That would allow us to eliminate an OpAmp and do the entire thing with a single dual OpAmp.
- [x] please annotate the schematic with the output filter characteristics (number of poles and 3dB bandwidth)
- [x] please can we increase the output filter 3dB BW to 500kHz. We should be able to run this board at 2MSPS update rate, so 330kHz is actually a bit slow (apologies, this is my fault for putting this too low in the original specification).
- [x] the johnson noise of the 5k resistors used on the DAC outputs is actually quite large. Is there a convenient resistor ladder with slightly lower impedance that we can use? If there isn't a convenient resistor ladder with slightly lower impedance then don't worry, these values will be ok.
- [x] please make sure to add test points on supply lines important digital signals (as we did in other projects)
- [x] power sequencing looks good: microprocessor won't drive the DAC SPI lines until after the DAC's 5V supply is on; DAC output is high impedance so won't damage the OpAmps; reference has proper power sequencing. So I don't think there will be any issues here!

Some context here:
- Our group will likely need Fastino at some point in the next year or so
- I'd like to use exactly the same DAC and AFE for Fastino that we use here (without any modification) if possible so that we can kill two birds with one stone
- For Fastino, we will need a lot of channels so space, cost, power consumption, cross-talk are all quite important. That's why I want to minimise the number of OpAmps and find a decent way of doing the reference buffering

**Comment by hartytp** [comment]
NB I'm not reviewing the power supplies section very carefully since this is copied from other projects which have already been through several design reviews. From a brief look, it looks good though!

**Comment by hartytp** [comment]
Nit pick on the PSU page: in the sequencing annotation, it should be "reference" or "ref" not "adc ref" since it's used for the adc and dac.

**Comment by hartytp** [comment]
- If we removed one MMCX from the FP would there be room for the micro USB? I suspect that most users would prefer USB on the FP to the 2xMMCX.
- Have you checked that there is enough clearance between the SMAs and LEDs/MMCX connectors, even when the isolating washers are mounted?

**Comment by hartytp** [comment]
- [ ] The RC components on the first page are missing values
- [ ] once everything else is finished, let's have a rough power budget (useful for specifying an appropriate PoE switch to drive several boards).
- [ ] IC24 missing a PN on the schematic
- It would be good to get feedback from other potential users (@cjbe @dtcallcock @dhslichter @jordens et al) about the pin allocation on the digital/EEM connectors. Is everyone happy with the peripherals those are connected to?

**Comment by hartytp** [comment]
@gkasprow I'm not going to check over the microprocessor/ethernet/USB parts of the schematic. I'm not the right person to review that.

**Comment by gkasprow** [comment]
We can use another edge hang  (in-line) type of USB connector and fit in under Ethernet jack.
![obraz](https://user-images.githubusercontent.com/4325054/48672563-b8be2d80-eb37-11e8-83ea-9c9bfa7d4104.png)
![obraz](https://user-images.githubusercontent.com/4325054/48672568-da1f1980-eb37-11e8-9882-7f84dc0f79d6.png)

**Comment by hartytp** [comment]
ADC:
- [ ] Like the DAC, the Iref is too large to use with a 10R filter resistor. On top of that, the capacitance could destabilise an OpAmp if there is no isolation resistor. We need to find a better reference buffer solution for this board.
- Are the LVCMOS buffers for the ADC just there to convert the 3V3 microprocessor levels to 2V5? (I'm assuming these have 3V3 tolerant inputs even when powered from 2V5?). 
- NB the ADC filter we have is quite a bit more aggressive than the one on the ADC data sheet. IIRC the bandwidth from this circuit is still fine, so that's fine IMHO (gives better rejection of high-frequency noise).

**Comment by hartytp** [comment]
Also, note that the ADCs draw quite a high reference current, but there is no buffer to isolate them. That could lead to some channel to channel and ADC to DAC cross-talk. Not sure if that's an issue, but worth a bit of thought.

**Comment by gkasprow** [comment]
The LVCMOS buffers are mainly to cut the noise that is propagated from CPU over the digital lines and couple with analog signals. If you want true 16bit performance it's a must.
Level conversion is another story.

**Comment by hartytp** [comment]
The 4.096V reference should usually be bypassed with 10uF, rather than 2uF (in a sensible case size to ensure that the capacitance doesn't drop when the bias voltage is applied).

**Comment by hartytp** [comment]
> The LVCMOS buffers are mainly to cut the noise that is propagated from CPU over the digital lines and couple with analog signals. If you want true 16bit performance it's a must.
Level conversion is another story.

Aah, good point.

**Comment by hartytp** [comment]
- [ ] axuiliary ADC and DAC filters are a bit aggressive. Let's go for a 1MHz pole frequency on both (users can add a more aggressive filter on the IDC to BNC board if they wish).
- [ ] please annotate the pole frequency on the auxiliary ADC and DAC schematic page

**Comment by hartytp** [comment]
> We can use another edge hang (in-line) type of USB connector and fit in under Ethernet jack.

If you think that will be okay mechanically then it's fine by me. i.e. will that be okay with the FP? will it be reasonably mechanically robust? will it be a real pain to insert the USB and ethernet at the same time? If all that's okay then it seems like a very good solution.

**Comment by hartytp** [comment]
@gkasprow that's my review complete. Thanks for a really good initial version.

The only big-is thing here is deciding what we want to do about the reference buffer. Other than that it's just a few small things...

**Comment by hartytp** [comment]
> The LVCMOS buffers are mainly to cut the noise that is propagated from CPU over the digital lines and couple with analog signals. If you want true 16bit performance it's a must.

If they're essential for the ADC, why aren't they essential for the DACs?

**Comment by gkasprow** [comment]
ADC is more sensitive to noise due to SAR architecture. Small interference can  take a few LSBs easily.
DACs are just current sources or resistor arrays so the noise does not cause such conversion errors. The input noise can only propagate to the output. I had such issues only with ADCs so far. It is caused by ground bouncing of the silicon chip.

**Comment by hartytp** [comment]
ok, makes sense. I didn't know that.

**Comment by hartytp** [comment]
One other thing:
- [ ] will there be an issue with using the same 2V reference for the microprocessor that we use for the ADC? It's okay to share that between the ADC channels, since it drives a high-impedance input. If the microprocessor draws quite a bit of current (and is quite noisy) then won't there be nasty cross-talk? Wouldn't it be better to either add a really cheap 2V5 reference right by the microprocessor, or at least add a reference buffer to isolate the microprocessor?

**Comment by gkasprow** [comment]
Roughly 15 years ago I built scientific grade [CCD camera](http://www.fuw.edu.pl/~gkasprow/usb_&_gbe_cameras.htm) for my master thesis. I used 16bit 10MS/s ADC. First prototype that used DRAM chips was working fine, but had higher noise (~40e-). Second one (with SDRAM, PCI and GB Ethernet) was much better with readout noise below 10e-. And to my frustration, every image had same pattern applied. It was roughly 10LSB pattern, same on every image. I tried many things - separated ADC from FPGA board, supplied with batteries, placed in metal cage. After hours of fight I noticed that only digital lines and GND are connecting the two. So I placed HCT244 buffer. The pattern disappeared and overall noise was greatly reduced to just 2 LSBs RMS including front end. It was even smaller than in datasheet. Their reference design did not use such buffering :D AFAIR I used AD9826 chip. The pattern was caused by SDRAM controller that went through exactly same sequence during every image acquisition that caused power supply bouncing. This signal went through digital lines to the ADC and coupled with analogue lines via internal capacitances and also impedance of ground  bondings.

**Comment by gkasprow** [comment]
Since that time I always place buffer between ADC and readout IC :)

**Comment by hartytp** [comment]
ouch.

**Comment by hartytp** [comment]
One other question: does one have to place a jumper to flash new firmware? If so, please can we add an annotation about that. Is there any way around that to make loading new firmware remotely (without having to add then remove a jumper) possible?

**Comment by gkasprow** [comment]
Firmware update can be triggered by software command.

**Comment by hartytp** [comment]
okay, good.

**Comment by hartytp** [comment]
One other slight nit pick:
- [ ] let's call them DI0 and DI1, not "adc trigger" and "dac trigger" (they can be used for different things in fw so let's keep the names general)

**Comment by hartytp** [comment]
@gkasprow so, the plan for the GPIO/DIO header is that we have 16 single-ended signals. If this is used with the IDC to BNC board to break them out then we must set 8 of those signals to LOW in firmware, and use the other 8 as GPIO. Are you happy that this plan will work okay?

**Comment by gkasprow** [comment]
@hartytp I' not sure if grounding BNC with CPU pin is good idea. It will damage CPU quickly with first serious EMI from i.e. nearby thunderstorm.

**Comment by hartytp** [comment]
@gkasprow ack. That's why I asked.

So, with the current arrangement the digital header is not compatible with the IDC BNC board, because that assumes that there are 8 differential signals on each header.

I can see a few options options (if you have a better solution let me know!):
1. Create a new revision of IDC_BNC that has jumpers that allow one to switch the BNC grounding between circuit ground and the _N half of the differential pair
2. Split the GPIO across two digital headers, each of which only has 8 signals.
3. If there are enough pins, then we can have two digital headers. One header has 8 single-ended signals and is designed to be compatible with the IDC_BNC board. No need for SPI/I2C or anything fancy here. Stick with GPIO and maybe a timer output. The second header has more digital signals with a mixture of GPIO/SPI/I2C etc. The second header is designed for connecting to custom AFE modules and is not compatible with the IDC_BNC board (it should use a different header so they cannot be mixed up by mistake). We should have clear annotations on the schematic labelling which headers are intended to be used with IDC_BNC.

If there is room then 3 sounds like the best option to me. What do you think?

**Comment by hartytp** [comment]
- [ ] Let's also make sure to expose the main 12V supply on one of the board-board headers (probably on the high-density digital one) using a few pins to give 3A or so. The use case I have in mind for this would be putting a 5A SMPS on an AFE mezzanine to create a compact ultra-low noise current source for coils etc.

**Comment by hartytp** [comment]
@gkasprow what do you think we should do about the reference buffers?

Looking at the LTC6655 data sheet, the voltage reference regulation starts to degrade after 1mA. Each DAC draws up to 0.55mA, and the two ADCs draw something comparable. By 2mA, the error due to the reference alone is something like 20ppm, so we will see a slight degradation of INL/DNL and cross-talk.

For Fastino with >=8 DACs per board, we definitely cannot drive the DACs directly from the reference. So, finding a good way of doing the reference buffering is going to be important. What's your preferred solution to this? Do we need to pick an OpAmp and design an RC compensation network to ensure both stability and low output impedance at all frequencies? The latter is doable, but will take some time in spice (unless you already have a working design for this?)

**Comment by dtcallcock** [comment]
> finding a good way of doing the reference buffering is going to be important. 

What about the circuit examples in the [datasheet](https://www.analog.com/media/en/technical-documentation/data-sheets/6655ff.pdf)?

**Comment by hartytp** [comment]
The LT3042 is a lovely LDO, which I've used as a ref buffer before. Load regulation is 0.5mV (max) for a change in load between 1mA and 200mA. Assume that's roughly linear (no idea how valid that is) and it gives approx 2.5mOhm output impedance.

Assume peak-peak change in ref current is 500uA*num_comverters. This gives 1uV/converter. For 4.096V reference, 1LSB=62uV, so that should allow us to drive up to 62 converters from a single reference buffer while still keeping the worst-case cross-talk below 1LSB pk-pk.

However, that's only at DC. Once we start adding 10uF per converter, what happens to the ref buffer BW? The concern would be if a frequency band appears where the reference is running out of bandwidth (and hence its output impedance rises), but the decoupling capacitors aren't low enough impedance yet (the DACs require something like 50mOhm max). That's not something that can be guaranteed from the specifications, and would need testing. That's if the thing is even stable with that much output capacitance, which isn't clear to me from the data sheet. For a few DACs and ADCs it's probably fine, but it would need testing. This path doesn't seem crazy to me and might be the best option for this design.

We could obviously use one LT3042 per converter, but that would be very expensive and bulky.

As to the other circuits on the LTC6655 data sheet (essentially just an emitter-follower), they're less likely to work. One would need a noise filter to reduce the 40nV/rtHz reference noise, while still maintaining enough BW to keep the output impedance low at higher frequencies. That would probably require an OpAmp circuit to drive the BJT, which would require careful design.

The other option is to pick a cheap OpAmp in a small package and design a compensation network to make it work. Make it cheap, compact and low enough power that we can use a separate buffer per channel. That has the advantage of providing excellent isolation between the channels.

**Comment by hartytp** [comment]
My guess is that @gkasprow has done this many times before and already has a preferred solution.

**Comment by hartytp** [comment]
> How about using MMCX?

If you're going to use coax then might as well just use the FP SMAs IMO. The internal MMCXs are okay, but can be a bit of a pain to route with all those IDCs.

My primary use-case for the "main analog" header is for plugging directly into custom AFEs like a current stabilization circuit, which screw directly to stabilizer and use the headers for board to board connections. That seems much nicer than using coax for board to board stuff.

**Comment by dtcallcock** [comment]
> My primary use-case for the "main analog" header is for plugging directly into custom AFEs like a current stabilization circuit, which screw directly to stabilizer and use the headers for board to board connections. 

Ah, I had forgot that you wanted to do that.

**Comment by hartytp** [comment]
@gkasprow NB when you do the routing: if we connect more than one converter to the same ref buffer then we must make sure to use a start layout for the reference connections to minimise the cross-talk due to trace impedance.

**Comment by gkasprow** [comment]
The original circuit that converts unipolar to bipolar output relies on internal DAC resistors.
We can combine gain stage with shifter circuit but have to connect external reference voltage to not include precision resistors in each channel. So we can generate 4/3 *4.096V reference and use the circuit below:
![obraz](https://user-images.githubusercontent.com/4325054/48972578-8f1c6f00-f02d-11e8-9870-5f5c20f06ec5.png)

**Comment by gkasprow** [comment]
edit: reference value

**Comment by gkasprow** [comment]
3.413V is 4.096 * 5/6

**Comment by gkasprow** [comment]
This can be achieved using 5 + 5 + 10 + 10 resistor, resistor ratio of 5:25 gives  1/6 and 5/6 which does the job.

**Comment by gkasprow** [comment]
Buffering references is challenging. I can drive 10uF with 30Ohm series resistor and still maintain DC accuracy, but it requires additional feedback loop (1 resistor + 1 capacitor).
Another idea is to use op-amp with external compensation that can drive indefinite capacitance, but I don't know any with very low offset (they typically have a few mV of offset)
Yet another idea is to use precise opamp in every DAC channel as we do for Stabilizer.
For Fastino it would increase the cost too much. We can also try to isolate the opamp with LC filter, but it is tricky.

**Comment by gkasprow** [comment]
The question is how big the isolation resistor can be. AD5542 defines in fig 17 that current changes between 100uA up to nearly 200uA. So our buffer must keep ref stability below 1 LSB, in this case 62uV 4.096/2^16). Iref change is 100uA, so maximal Rs = 0.62Ohm. This is also the maximum value of reference buffer. If we want to drive 32 channels from same reference, our buffer must withstand max and min load for all buffers. So its output resistance must be lower than 0.62Ohm/32 =  19mOhm. Since the DACs change the reference current rapidly, the buffer must react quickly and not oscillate. 
The LT3042 you proposed could work, but with tantalum capacitors which have much higher capacity.
Analog Devices in [CN0079 note](https://www.analog.com/media/en/technical-documentation/application-notes/an-1525.pdf) presents idea of using individual AD8628 with feedback taken via REFS.
Such solution would increase the channel cost by roughly 2 EUR. On the other hand,  cost of AD5542 (10EUR), AD8620 (11EUR) and AORN2-1AT3 (2.7EUR) is already 24EUR so additional reference buffer is less than 10%. AD8628 is available in TSOT so we can fit them easily and use  it for distribution of other references due to very low offset (1uV)

**Comment by gkasprow** [comment]
AD8628 in SOT-23 costs 1.5EUR@100pcs

**Comment by gkasprow** [comment]
Another issue can be with delivery of 3.413V reference. But here situation is easier. Every channel consumes +68uA / -340uA of current, but the impedance is always 10k. Even with 32 channels we still have 312Ohm.
The reference buffer must maintain -2.1mA / +10.88mA but every opamp can do that. Of course at high frequency its performance will degrade and the crosstalk may get worse. We must make sure that its BW is much higher than amplifier BW.

**Comment by hartytp** [comment]
Regarding the reference buffer, I can see a few options:
1. LT3042. Maybe limit to max 8 converters per reference buffer to limit cross-talk. For Fastino we'd need quite a few reference buffers. The DC output impedance of this buffer is extremely low, and it has a wide bandwidth when loaded with 4.7uF, allowing it to keep a low impedance across a broad frequency range. The biggest issue I see here is how well the reference buffer performs at high frequency when we load it with a lot of capacitance. However, since the buffer is quite fast, the best bet might be to place it physically near the converters, connecting it with short wide traces to keep the trace impedance low. We can then use something like 1uF capacitors on each reference, so the total capacitance on the LT3042 output is not too high.  
2. Add 1 OpAmp buffer per converter. Pick a fast OpAmp with good current drive so that no capacitor is needed on the reference pin (i.e. we rely on feedback to keep the reference impedance low at all frequencies, rather than the standard approach of using capacitance). The big concern here is that while this will work well at DC, if the OpAmp isn't fast enough, this might lead to worse DAC glitches/AC non-linearity since the reference source impedance can rise at higher frequencies.
3. Use 1 OpAmp buffer per converter. Add the usual 10uF||100nF capacitors, add an isolation resistor + RC compensation network to the OpAmp to stabilize it. This would need careful simulation.

The challenge with any of these approaches is ensuring that the reference pins have see a low impedance across all frequencies.

@gkasprow which solution do you think we should go for. I'm tempted by (1). It's definitely the cheapest and should work well so long as we watch the load capacitance on the LT3042 as well as the trace impedances. What do you think?

**Comment by hartytp** [comment]
> Another issue can be with delivery of 3.413V reference. But here situation is easier. Every channel consumes +68uA / -340uA of current, but the impedance is always 10k. Even with 32 channels we still have 312Ohm.

It's not really easier here. While there is a 10k resistor, the effective impedance changes significantly (see how much the current changes by). We still have a large, non-static current draw on this reference, so it's a hard problem to solve.

As always, it's easy at DC, but avoiding glitches/cross-talk/poor AC INL is the tricky part. Much better to avoid references driving summing junctions where possible

**Comment by gkasprow** [comment]
That's why I wrote that we need fast opamp. Glithes are not an issue here because are filtered by first stage opamp. Let's try with combined filter and gain stage.

**Comment by hartytp** [comment]
> Let's try with combined filter and gain stage.

Sounds like a good plan. Can you do a quick spice sim of that?

**Comment by hartytp** [comment]
> Can you do a quick spice sim of that?

Otherwise, if you post a schematic then I can do a SPICE work up some time this coming week.

**Comment by gkasprow** [comment]
this could work
![obraz](https://user-images.githubusercontent.com/4325054/48978002-7485dc80-f0a5-11e8-834a-a8674c44763a.png)

quick simulation of second stage

![obraz](https://user-images.githubusercontent.com/4325054/48978004-810a3500-f0a5-11e8-9886-453630432a8a.png)

**Comment by gkasprow** [comment]
The only weak point is output resistance of the DAC that has 20% tolerance. I added 10k to reduce its influence, and made the first stage slightly above 500kHz. It's worth to simulate how it effects the BW and also performance with real model of AD8620.

**Comment by gkasprow** [comment]
Simulation with ADIsimPE
![obraz](https://user-images.githubusercontent.com/4325054/48978278-80739d80-f0a9-11e8-9a13-eae6a411b38d.png)

**Comment by gkasprow** [comment]
This 300pF capacitor looks surprisingly large. so there is sth wrong with design tool I used.

**Comment by gkasprow** [comment]
I found where the problem lies.
The AD8620 has 15pf of input capacitance. And filter wizards do not take it into account. The DAC output resistance with input capacitance already makes 500kHz LP filter. So when I added 10k resistor to mitigate 20% tolerance of internal resistance, I cut the BW to 377kHz.

**Comment by gkasprow** [comment]
Edit - I used wrong resistances at the DAC level shifter.
When I change them from 10k do 28k as DAC has, opamp input capacitance increases the BW up to 700kHz! So we have to add compensation capacitance in the first stage as well
![obraz](https://user-images.githubusercontent.com/4325054/48979133-b1a69a80-f0b6-11e8-97a4-2ec3e98f43f1.png)

**Comment by gkasprow** [comment]
In this way we achieve:
![obraz](https://user-images.githubusercontent.com/4325054/48979141-c125e380-f0b6-11e8-9d92-68881ced8517.png)

**Comment by gkasprow** [comment]
Let's analyse the first stage only. Without compensation we get:
![obraz](https://user-images.githubusercontent.com/4325054/48979188-3a253b00-f0b7-11e8-933e-1c49ea252215.png)

The input capacitance on negative terminal of opamp is responsible for such peaking.

**Comment by gkasprow** [comment]
With 15pF compensation capacitor we get smooth response at 1.9MHz. So let's add some resistance to the DAC output and also some capacitance to the opamp input to make the filter less sensitive to the DAC output resistance variation. In semiconductor process the capacitances are controlled with high accuracy, but it's hard to control resistances. But one can control resistance proportions, that's why the Rinv and Rfb have 20% tollerance but very good matchning.
So let's add 10k and 5p to get 500kHz first order filter.
![obraz](https://user-images.githubusercontent.com/4325054/48979297-064b1500-f0b9-11e8-8e4e-049fe48c3a82.png)

**Comment by gkasprow** [comment]
Of course when we combine 2 filters, we will get 500kHz cut-off at -6dB so individual BW must be trimmed.
And have to compensate for op-amp capacitances otherwise will get peaking above 1MHz.
![obraz](https://user-images.githubusercontent.com/4325054/48979548-faf9e880-f0bc-11e8-8b71-68ed43749775.png)
![obraz](https://user-images.githubusercontent.com/4325054/48979552-06e5aa80-f0bd-11e8-9964-3d3ef82556cb.png)

**Comment by hartytp** [comment]
@gkasprow thanks for doing that! That response curve looks really nice. The plot with component tolerances taken into account looks good too. I'm happy with that design for the filter. Good work.

**Comment by hartytp** [comment]
NB the compensation capacitor on the first OpAmp is included in the reference design for the LTC2642 (see figure 1 in the data sheet).

**Comment by gkasprow** [comment]
well, these days it's hard to invent anything :)

**Comment by hartytp** [comment]
@gkasprow one final thing about this: is it worth doing a simulation of the step response/settling time to check that the OpAmp slew rate isn't a limitation when used in this filter configuration.

Other than that, all we need to do is make a decision about the reference buffer and I think the major design decisions are complete.

**Comment by gkasprow** [comment]
If we use local reference buffer like AD8628, than the issue would be solved.
![obraz](https://user-images.githubusercontent.com/4325054/48983946-9c9d2c00-f0f5-11e8-894a-a64651434eaa.png)

**Comment by hartytp** [comment]
@gkasprow maybe I'm missing something but I'm not sure that the design you've sketched out will work.

We want to run these DACs close to the 3MSPS max sample rate so the impedance at the reference pin must be low (<50mOhm) up to at least a few MHz. Look at figure 20 on the data sheet. With Av=1 the OpAmp output impedance is close to 60Ohm at 2MHz! The feedback will not be able to compensate for the voltage drop across the OpAmp's output impedance because the component has a GBP of 2MHz, so there is not sufficient gain above 10kHz or so.

I believe that this circuit will work if the DAC is used close to DC, but not in a high sample rate application like ours.

Or, am I missing something?

**Comment by hartytp** [comment]
> AD5542 defines in fig 17 that current changes between 100uA up to nearly 200uA. So our buffer must keep ref stability below 1 LSB, in this case 62uV 4.096/2^16). Iref change is 100uA, so maximal Rs = 0.62Ohm. This is also the maximum value of reference buffer. If we want to drive 32 channels from same reference, our buffer must withstand max and min load for all buffers. So its output resistance must be lower than 0.62Ohm/32 = 19mOhm. Since the DACs change the reference current rapidly, the buffer must react quickly and not oscillate.

I think you mean figure 18?

Note that this is for a 2V5 reference. For 4.096V the current is higher. In bipolar operation the minimum reference impedance is 7.5K, so the maximum reference current is 550uA. I do not think that driving 32 channels from a single reference buffer will ever work. 4 or even 8 channels per reference buffer could work depending on the buffer.

**Comment by gkasprow** [comment]
I mean [ design note](https://www.analog.com/media/en/technical-documentation/application-notes/an-1525.pdf)

**Comment by hartytp** [comment]
Large signal gain plots look good! Okay, I'm happy with that filter design. :)

**Comment by gkasprow** [comment]
What about such design?
10u cap takes care about high frequency components while opamp keeps DC votage stable.
![obraz](https://user-images.githubusercontent.com/4325054/48984235-b3de1880-f0f9-11e8-98a4-4d80bb81030b.png)

**Comment by gkasprow** [comment]
let's simulate it

**Comment by hartytp** [comment]
@gkasprow I've used that kind of design before, but it still needs some simulation to ensure that the output impedance is low at all frequencies (we don't want bands where the reference driving impedance peaks) and that the circuit is stable.

Let's do some really crude back-of-the envelope maths to see if this is likely to work...

How low does the impedance need to be? 4.096/2^16 = 62.5uV. So for 550uA max reference current, a resistance of 0.1Ohm gives a change in the reference value of 1LSB. So, let's assume that the reference driving impedance must be less than 0.1Ohm up to a few MHz.

10uF reaches 0.1 Ohm at 160kHz. So the OpAmp must ensure that the feedback compensates for the voltage drop across the 10Ohm resistor (as well as any OpAmp output impedance) up to at least 160KHz. Since we need to reduce the output resistance by a factor of 100, we require an output open-loop gain of approx 40dB up to 160kHz. The OpAmp output impedance is only 30dB at 100kHz, so I suspect that this particular OpAmp is not fast enough for this application. 

To have 40dB of open loop gain at 160kHz, we need a min OpAmp GBP of 16MHz (assuming dominant pole compensation for the OpAmp).

The pole frequency in your compensation network is at 7MHz. This is above the OpAmp's unity gain cross-over so I suspect that it will not be sufficient to stabilise the circuit. Note that for this application it is very important that the OpAmp output does not ring, so we need to make the circuit nice and stable.

> let's simulate it

Yes, a proper simulation would be a really good idea. However, my crude maths above makes me think that it will not work with this OpAmp...

**Comment by gkasprow** [comment]
The circuit
![obraz](https://user-images.githubusercontent.com/4325054/48984390-51861780-f0fb-11e8-8a72-c3d9a17cbbe1.png)

current generator injects +/- 100uA  1MHz square wave. And we get:
![obraz](https://user-images.githubusercontent.com/4325054/48984405-78dce480-f0fb-11e8-8e1d-d18b125b4a0d.png)

Ripples are roughly 10uV ad are modulated by 70us sinewave
![obraz](https://user-images.githubusercontent.com/4325054/48984431-c5c0bb00-f0fb-11e8-940a-e54379f15595.png)

**Comment by hartytp** [comment]
> current generator injects +/- 100uA 1MHz square wave.

I think this should be 550uA for a 4.096 reference voltage, since the min specified reference impedance is 7.5k

**Comment by hartytp** [comment]
hmmm...@gkasprow I'm surprised that this circuit is stable. I would have expected that the 220Ohm||100pF compensation network wouldn't do anything until around 7MHz, which is beyond the OpAmp unit-gain cross-over. So the circuit should oscillate. But maybe I'm missing something?

On top of that, I'm not sure that a 1MHz sure wave is a completely fair test. The circuit will work really well at 1MHz since the OpAmp does nothing and the response is essentially just the response of the capacitor 20mOhm ESR. The toughest region will be where the gain of the OpAmp + compensation network starts to decrease, so that the 10Ohm isolation resistor is not fully bootstrapped.

I think we need to look at the output impedance as a function of frequency and check that it is low at all frequencies, without any peaking or instability.

**Comment by gkasprow** [comment]
Typical ESR for 10uF 0603 capacitor is 5-10mOhm at 1MHz. LEt's asume 10mOhm.
I had to tune loop parameters to make it stable.
![obraz](https://user-images.githubusercontent.com/4325054/48984604-cd815f00-f0fd-11e8-9b67-daa62550853e.png)

**Comment by gkasprow** [comment]
The ripples are 0 / 550uA

**Comment by gkasprow** [comment]
Actually the circuit was not stable. It was oscillating with 70us period:). Now it is stable.

**Comment by hartytp** [comment]
> https://github.com/sinara-hw/Stabilizer/issues/2#issuecomment-441472942

I'm not sure I understand what this plot shows. Does this mean that when you turn the reference current from 0uA to 550uA there is a transient change in the reference voltage of 2mV=35LSB, which takes something like 1ms to settle??

**Comment by gkasprow** [comment]
I modulated the output with additional 5kHz, 100uA square wave
Here we see opamp response
![obraz](https://user-images.githubusercontent.com/4325054/48984636-7760eb80-f0fe-11e8-99f2-09abb9438bc6.png)

**Comment by hartytp** [comment]
> I modulated the output with additional 5kHz, 100uA square wave
> Here we see opamp response

So for 100uA at 5kHz, that gives 500uA pk-pk, which is nearly 10LSB. The full 500uA reference current changes we can expect, that would give a really quite large change in the reference voltage.

I think this is an inescapable issue with this OpAmp; it's just not fast enough. As I noted above, to make the circuit work, the OpAmp must bootstrap out the isolation resistance, reducing the effective resistance by a factor of 100 at 160kHz. As a result, the GBP must be around 16MHz or the circuit cannot do the job. So, I do not believe that any compensation network will make this circuit work without changing the OpAmp for a faster one...

**Comment by gkasprow** [comment]
I replaced the opamp with AD8620, tuned the loop to get fast response and ripples are still 370uV

**Comment by hartytp** [comment]
> I replaced the opamp with AD8620, tuned the loop to get fast response and ripples are still 370uV

What's the circuit?

**Comment by gkasprow** [comment]
I used AD8620 as follower. No RC components, just opamps and supplies.
And we get 
![obraz](https://user-images.githubusercontent.com/4325054/48984737-f276d180-f0ff-11e8-836b-ae51605f87fc.png)

**Comment by hartytp** [comment]
> I used AD8620 as follower. No RC components, just opamps and supplies.

Aah, I wouldn't expect that to work. From data sheet figure 30 Zout is at least an Ohm or two by 1MHz, which is at least an order of magnitude too high for this application. The isolation resistor + 10uF capacitor + compensation network is always likely to be required when using this kind of buffer at MSPS sample rates AFAICT.

**Comment by gkasprow** [comment]
The circuit was the same, R was 1k, C was 1n.
We will always get some voltage peaks. The opamp has limited feedback time.

**Comment by hartytp** [comment]
Cool, there is a spice model. Let's have a look then. Try with a 4.7uF output capacitor.

**Comment by gkasprow** [comment]
sure, There is full design. I will have results in a minute.

**Comment by hartytp** [comment]
also, cool app note in the Jim Williams style https://www.analog.com/media/en/technical-documentation/application-notes/an159fa.pdf

**Comment by gkasprow** [comment]
Oh, I remember it. This original metal box :)

**Comment by gkasprow** [comment]
OK, so with LT3042 and 500uA amplitude I get:
![obraz](https://user-images.githubusercontent.com/4325054/48985118-d0cc1900-f104-11e8-9178-47a90f2ba358.png)
![obraz](https://user-images.githubusercontent.com/4325054/48985141-0cff7980-f105-11e8-9620-24a8757e12bc.png)

**Comment by gkasprow** [comment]
I lowered SET capacitance to 1nF to make simulation faster. And pulses are 0 - 550uA.
Now I get 20uV ripples.

**Comment by gkasprow** [comment]
![obraz](https://user-images.githubusercontent.com/4325054/48985256-eaba2b80-f105-11e8-9b92-2098eb5c71ba.png)

**Comment by gkasprow** [comment]
I placed 10u, 10mOhm cap
![obraz](https://user-images.githubusercontent.com/4325054/48985267-11786200-f106-11e8-93c7-966f3731e9c4.png)

**Comment by hartytp** [comment]
> I lowered SET capacitance to 1nF to make simulation faster. And pulses are 0 - 550uA.
Now I get 20uV ripples.

Nice! 0-550uA is a bit pessimistic, since figure 18 on the data sheet suggests that the change in the reference current is only about 50%. So, maybe 200uA pk-pk is more accurate.

So, this suggests that we can drive something like 4 data converters from a single LT3042 while still keeping cross-talk/glitches/etc at or below 1 LSB pk-pk

**Comment by hartytp** [comment]
@gkasprow to be sure this works, we should simulate for several frequencies of square waves (10Hz, 30Hz, 100Hz, 300Hz, 1kHz, 3kHz, 1MHz, 3MHz) and check what the worst-case response is. If we always see 20uV pk-pk for a 0-500uA square wave then this seems like it will work beautifully.

**Comment by gkasprow** [comment]
This is exactly what I do right now.

**Comment by gkasprow** [comment]
3 plots
1MHz, 550uA
100kHz, 200uA
10kHz, 100uA
![obraz](https://user-images.githubusercontent.com/4325054/48985337-a7ac8800-f106-11e8-8357-667f7561ac6a.png)

**Comment by hartytp** [comment]
> So, this suggests that we can drive something like 4 data converters from a single LT3042 while still keeping cross-talk/glitches/etc at or below 1 LSB pk-pk

Hmm...if we want to use one ref buffer with multiple converters then we have to think carefully about capacitance. Do we put, say, 2uF by each converter and 2uF by the LDO to give 10uF total?

**Comment by gkasprow** [comment]
ah, the  10kHz pulse was 16mA, to simulate all channels update from 0 to max

**Comment by gkasprow** [comment]
so we cannot really drive all DACs from single source

**Comment by hartytp** [comment]
> ah, the 10kHz pulse was 16mA, to simulate all channels update from 0 to max

So, 16mA gives about a 2mV glitch at 10kHz? So, 200uA gives something like 20uV.

> so we cannot really drive all DACs from single source

I wouldn't expect that (32 converters from one buffer is just sadistic). But, 4 should be okay. £2.60 split 4 ways is pretty reasonable IMHO...

**Comment by gkasprow** [comment]
@hartytp true, this looks like good solution.
I combined sources:
1MHz, 100k, 10k, 100Hz, all 200uA
![obraz](https://user-images.githubusercontent.com/4325054/48985414-8b5d1b00-f107-11e8-9c26-03c6cb173162.png)

**Comment by hartytp** [comment]
Nice! 

> I combined sources:
> 1MHz, 100k, 10k, 100Hz, all 200uA

That's actually a bit pessimistic (which is good) since the peak combined fluctuations from all those sources are >500uA.

So, the pk-pk ripple is something like 3*7uV=21uV. Nice!

**Comment by gkasprow** [comment]
The last issue is LT3042 buffer offset voltage which is +/- 1mV. Do we care about it or shall we add option to trim it?

**Comment by hartytp** [comment]
So, the two remaining questions:
1. How many converters do we drive from each LT3042?
2. How much capacitance do we add to the output?

Let's assume we want to make the worst-case pk-pk fluctuations < 1LSB. No point aiming for 0.1LSB IMHO since that's just a waste of buffers.

Based on the numbers you posted, it seems that 1 LT3042 per 4 data converters should be about optimal.

Assuming you're happy with that, the final thing to do is to simulate the load of 4 DACs and tweak the capacitance to minimise the combined ripple from all those sources.

**Comment by gkasprow** [comment]
I got it. It should work. I will simulate it.

**Comment by gkasprow** [comment]
it's quite possible that such number of LTCs will need some individual load to make all feedback loops active.

**Comment by hartytp** [comment]
hmm...if we do parallel them then we need to be really careful to compensate the voltage drop across the 50mOhm ballast resistors. There is a circuit for that in the data sheet, but it's a little complex and I'm not sure how well it works with fast dynamic loads.

**Comment by gkasprow** [comment]
maybe it's easier to use single reference and compensate the offset with DC opamp?

**Comment by hartytp** [comment]
Maybe a better option for removing the offsets is:
- do not parallel the LT3042 buffers. Just use one independent buffer per 4 data converters as discussed before
- for each LT3042, add a cheap low power OpAmp (nice small package like SOT23-5)
- do some DC feedback onto the set pin using the OpAmp

**Comment by gkasprow** [comment]
You're reading my mind :)

**Comment by gkasprow** [comment]
AD8628 with 1uV offset should do the job.

**Comment by hartytp** [comment]
> maybe it's easier to use single reference and compensate the offset with DC opamp?

You beat me to it :)

I think it's worth adding an OpAmp to trim out the LT3042 offset. +-1mV=+-15LSB which is an order of magnitude worse than the DAC itself. The OpAmp we use needs to be low noise, but can still be quite cheap, and low power, so it seems worth it.

**Comment by hartytp** [comment]
Ok, so we just need to optimise the output capacitance for the LT3042 regulators and the hard parts of the design are finished :) good!

**Comment by hartytp** [comment]
@gkasprow NB assuming we drive 4 DACs from the LT3042 instead of 8 then we would need 8 LT3042s on Fastino. However, Iset is 100uA per buffer. A LTC6655 would struggle to sink that much current, so we will need a reference buffer to drive the reference buffers!

**Comment by hartytp** [comment]
(or just reduce the number of DACs on Fastino, but let's worry about that some other day)

**Comment by gkasprow** [comment]
We can drain this current to GND with resistor in every LDO and then connect LTC6655 with low ohm resistor. Anyway we will need some current summing junction to compensate for offsets.

**Comment by hartytp** [comment]
@gkasprow I feel a bit bad using the AD8620 since it's quite expensive (£5 per channel), so I had another look for OpAmps.

I think this search captures what we need: https://www.analog.com/en/parametricsearch/11070#/p300=15|5500&p2840=24|220&p4101=8.5e-10|15&p4124=0|1e-12&p4127=0.1|1&p4500=2e-14|2.5e-8&p4501=0.0000025|0.00025&p4502=15000000|4000000000&d=sel|4501|4500|4502|300|4115|4127|2687|5034|4101|4124|s3|s5|4108|2839|2840

It's hard to find a good amp since we need +-12V rails, low bias current/current noise (source impedances around 10k), low voltage noise, low offsets, high speed. There aren't many choices. I think the other options are all worse and about the same cost, so let's stick with the AD8620.

**Comment by gkasprow** [comment]
what about such reference buffer? 
![obraz](https://user-images.githubusercontent.com/4325054/48986187-0fb49b80-f112-11e8-99c7-f100de34b441.png)

ADA4528 has:
2.5uV offset  
noise 5.6 nV/√Hz at f = 1 kHz,  and 97 nV p-p at f = 0.1 Hz to 10 Hz

**Comment by gkasprow** [comment]
RC values need some optimisation.

**Comment by hartytp** [comment]
That looks like just what I had in mind. As you say, RC values need double checking/optimising. In particular, we should optimise the load capacitance (will want at least 1uF ceramic near each DAC, ideally a bit more).

We will also want an RC filter at the output of the LTC6655 to suppress the 40nV/rtHz from its output buffer.

Probably easiest to make R157 a 40.96k resistor (or nearest convenient value) as recommended in the data sheet.

**Comment by jordens** [comment]
We haven't had the funding or free time to invest much into the review of this. Just one thing to consider: IME having a way to peek the inputs using the µC ADCs and poke/modulate the outputs using the µC DACs would be very useful. Even though they are of limited resolution, their much better DMA and timer support and integration into the fabric enables efficient modulation and demodulation with much less overhead than the external ADC/DACs.

**Comment by hartytp** [comment]
@jordens ack on the lack of funding (although, hopefully if this turns out to be a useful design there will be funding to develop really nice FW to go with it).

> Just one thing to consider: IME having a way to peek the inputs using the µC ADCs and poke/modulate the outputs using the µC DACs would be very useful. Even though they are of limited resolution, their much better DMA and timer support and integration into the fabric enables efficient modulation and demodulation with much less overhead than the external ADC/DACs.

interesting point. I'll need to give that some thought. Some off the cuff comments:
- I want to make sure that the uProcessor ADCs don't inject noise on to the main ADC input. May need a separate driver/buffer amp to isolate them
- Simillarly for the DAC, I'd be worry about noise coupled onto the output from the microprocessor and also offset drifts in the uProcessor DAC
- need to check but we'd probably need an extra OpAmp in the AFE to sum the two DACs together.
- IIRC the microprocessor only have two DACs and we're currently routing them both to the aux header. So, there isn't a spare DAC right now.

**Comment by jordens** [comment]
It would be great if the firmware development and design could already provide input and improve the end result early, especially at the hardware spec and design phase.
Just use jumpers for isolation and DAC choice.

**Comment by hartytp** [comment]
@jordens yes, it would be great to have funding for the firmware side of this project already. Unfortunately that's lacking right now as far as I'm aware. It's not ideal, but what we have to work with for the time being.

I suspect that this design will also prove useful to other Sinara users in the long run, so it would also be great to have input from a larger pool of potential users.

**Comment by gkasprow** [comment]
STM32 can control CS line and use DMA to transfer data from external ADC and to external DAC.
There are a few CS control modes, some use I2S-style, some use Motorola style.
I used SPI DACs with DMA in STM32F4 family so far and there were always some issues. Finally I managed to make it working using some tricks like one that next transfer latches previous sample :) 
It's quite possible that we will need some glue logic to make such ADC and DAC working in DMA mode due to this CS line control.
We can mitigate it and implement in very first prototypes some tiny CPLD to be able to play with it. Later on it can be eliminated or replaced by discrete logic. Some clock from CPU which is synchronous with SPI would be needed as well. And this is my proffered approach. Either 2$ ICE40 or CoolRunner CPLD will do the job. The SPI mode in ADC looks tricky. I forgotten to write about it, but afer making CS low we have to wait before transferring data. So will probably need to work in I2S mode.

**Comment by gkasprow** [comment]
There are some limitations how SPI and DMA can be combined. So I'd route all SPI + some interesting IOs to small CPLD and then to ADCs and DACs.

**Comment by hartytp** [comment]
@gkasprow I'm a little wary of doing that since it adds extra complexity to the design (who's going to write the CPLD code) and one more chip to program.

Is DMA important for this kind of design? How well does controlling the DACs + ADCs from an interrupt loop work for reading/playing waveforms?

**Comment by hartytp** [comment]
@jordens to come back to your previous point, this is where I would really value feedback from other potential users. What do people imagine using this design for? Without knowing that, it's hard to optimise the hw or specify what the fw should do.

Currently, my top priority is to make something that will work well for magnetic field stabilisation (our immediate need). I'm trying to make the design general so it will be useful for a wide range of uses, but without anything in particular in mind.

**Comment by jordens** [comment]
The U Hannover folks should chime in soon.
I'd want to use it for:

* Wavelength modulation spectroscopy and modulation transfer spectroscopy at about 100 kHz modulation, 1 MHz ADC/DAC bandwidth, ~1.6 MHz ADC/DAC sample rate
* PDH and FM locks with external modulation/demodulation and added intelligence w.r.t. autolock, feature recognition
* Frequency offset locks with external f-to-V conversion
* Hänsch-Couillaud locks
* Current stabilization

**Comment by gkasprow** [comment]
@hartytp I can writ simple VHDL code that connects inputs with outputs. When necessary, I can add some logic that generates necessary signals. In next revision we will get rid of the chip.  Doing 1.6MS/S ADC and DAC service in the interrupt routine will overload the CPU.

**Comment by hartytp** [comment]
> @hartytp I can writ simple VHDL code that connects inputs with outputs. When necessary, I can add some logic that generates necessary signals. In next revision we will get rid of the chip

Okay, sounds good.

This is the kind of detail where more input from other users and people who've thought more about the firmware implementation would be really helpful!

**Comment by hartytp** [comment]
>  Doing 1.6MS/S ADC and DAC service in the interrupt routine will overload the CPU.

Really? I figured that with a 400MHz CPU and HW SPI cores, we'd be okay running feedback loops on the microprocessor at 1MSPS. What kind of sample rate do you think we'll be able to achieve for a simple PID loop implemented on the microprocessor?

**Comment by gkasprow** [comment]
IF you write the ISR routine in assembly, without putting registers on the stack and taking them back, this might work with such rate. I'm not sure if the CPU can access peripherals with such speed - it will probably insert wait states. And access to the FLASH also goes via cache, so it needs a lot of optimisation to run it in ISR. That's why DMA is far more optimal method.

**Comment by hartytp** [comment]
@gkasprow okay. In that case, putting a CPLD on the board to allow us to prototype DMA to the ADCs and DACs sounds like a good idea.

**Comment by jordens** [comment]
We do the sampling and modulation in DMA and timers at 1.6 MHz. The processing is in an ISR at 100 kHz and I would not expect much more (more than a factor of 5 maintaining a decent feature set) from a µC. I don't really see the point in rolling out an extra CPLD and related gateware/software components if the µC ADCs and DACs can do it just fine and are just a jumper away and have really good and flexible DMA/timer connectivity. For the case where the external ADCs and DACs are needed, I'd first look very deep into what's needed in terms of pin choice and CPU configuration to DMA/timer/trigger them properly without using a CPLD.

**Comment by hartytp** [comment]
> The processing is in an ISR at 100 kHz and I would not expect much more (more than a factor of 5 maintaining a decent feature set) from a µC

Depends a little on what one considers a "decent feature set". For moderately complex control loops only, but no modulation etc, I'd expect quite a bit more than 100kHz on a 400MHz uC.

@cjbe remind me what update rate you got when playing around with PICs.

**Comment by gkasprow** [comment]
LEt's place tiny CPLD like XC2C64 and bypass 0R resistors below.
So it will run without issues by default.
When needed, someone can burn configuration, desolder 0R resistors and push ADC/DACs to the limits.

**Comment by jordens** [comment]
Processing two channels, each 16 tap FIR, decimate, two biquad IIRs, a couple saturations, output, all in integer 16 resp. 32 bits around 400 cycles here. Leave a good 50% margin and you could get to 500 kHz on that µC with that feature set.

**Comment by hartytp** [comment]
@gkasprow looks good so long as you're sure that this won't impact the main ADC's noise levels by coupling noise from the microprocessor to the main ADC.

What is the current draw from the CPU ADC?

**Comment by hartytp** [comment]
Please can you make sure that there is some silk around R162 and R15 so they're easy to identify if they need to be removed to disconnect the CPU adc?

**Comment by gkasprow** [comment]
ADC current flows during charging sampling cap. So one must make sure that capacitor is large enough to not cause systematic drop during charging. We had such issue with rev1.0 of Sampler and had to redesign the input stage. I can make these resistors not mounted by default since probably not many users need it.

**Comment by gkasprow** [comment]
@hartytp I assigned ADC channels to the BNC_DIO connector in such way that can be used either single-ended or differential. User can connect 0-2V to every BNC or treat neighbouring BNCs as P and N signals of diff pair and apply 4Vpp diff signal. Of course we have 3 ADC channels instead of 6.
2 another diff ADCs are routed to the AFE as I wrote above.
@jordens to make use of DACs we have two options:
- make simple AFE mezzanine with dedicated front panel connectors.
- try to connect existing output stage using resistor options instead of original DACs. It needs  three sets of dual 0R resistors per channel: on output signal and also on the DC shifter amplifier, to centre the DAC output around 1.024V instead of 2.048V which is right now. So to change the signal routing the user would need to modify 12 resistors. I'm not sure if it is worth the effort and mezzanine is better option.

**Comment by gkasprow** [comment]
Anyway CPU DACs and ADCs would be easily available via BNC_IDC adapter in current board version.

**Comment by jordens** [comment]
@gkasprow I think for the DACs AC coupling the µC DACs onto the external precision DAC outputs with a knee at ~10 kHz would already enable what I want. That would also avoid the shifting.

The ADC solution looks fine to me. If you want to follow @hartytp suspicion that ther emight be noise coupled out of the µC ADC onto the external ADC inputs, then I'd very much prefer a 0R to be added/removed (between C150/C151 and the ADC so removing it would also unload the IDC26 connector ADC pins) instead of adding/removing R15/R162.

**Comment by jordens** [comment]
Exactly. Thanks!

**Comment by gkasprow** [comment]
@hartytp @jordens Before I do board routing, please have a look at schematics and placement
https://github.com/sinara-hw/Stabilizer/releases/tag/v0.9rc2
